packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -629,3 +656,157 @@ class AsyncGuidelinePainter extends CustomPainter {
+    int timestampMicros = _startingTimestamp(intervalWidth, microsPerInterval);
+    double lineX;
+    if (visible.left <= chartStartInset) {
+      lineX = chartStartInset - visible.left;
+    } else {
+      lineX =
+          intervalWidth - ((visible.left - chartStartInset) % intervalWidth);
+    }
+
+    while (lineX < constraints.maxWidth) {
+      _paintTimestamp(canvas, timestampMicros, intervalWidth, lineX);
+      _paintGridLine(canvas, lineX);
+      lineX += intervalWidth;
+      timestampMicros += microsPerInterval;
+    }
+  }
+
+  void _paintTimestamp(
+    Canvas canvas,
+    int timestampMicros,
+    double intervalWidth,
+    double lineX,
+  ) {
+    final timestampText = msText(
+      Duration(microseconds: timestampMicros),
+      fractionDigits: timestampMicros == 0 ? 1 : 3,
+    );
+    final textPainter = TextPainter(
+      text: TextSpan(
+        text: timestampText,
+        style: const TextStyle(color: timestampColor),
+      ),
+      textAlign: TextAlign.right,
+      textDirection: TextDirection.ltr,
+    )..layout(maxWidth: intervalWidth);
+
+    // TODO(kenz): figure out a way for the timestamps to scroll out of view
+    // smoothly instead of dropping off.
+    final xOffset = lineX - textPainter.width - timestampOffset;
+    if (xOffset > 0) {
+      textPainter.paint(canvas, Offset(xOffset, origin + 5.0));
+    }
+  }
+
+  void _paintGridLine(Canvas canvas, double lineX) {
+    canvas.drawLine(
+      Offset(lineX, origin),
+      Offset(lineX, constraints.maxHeight),
+      Paint()..color = gridLineColor,
+    );
+  }
+
+  double _intervalWidth() {
+    final log2ZoomLevel = log2(zoom);
+
+    final gridZoomFactor = math.pow(2, log2ZoomLevel);
+    final gridIntervalPx = baseGridIntervalPx / gridZoomFactor;
+
+    /// The physical pixel width of the grid interval at [zoom].
+    return gridIntervalPx * zoom;
+  }
+
+  int _microsPerInterval(double intervalWidth) {
+    final contentWidth = flameChartWidth - chartStartInset - chartEndInset;
+    final numCompleteIntervals =
+        (flameChartWidth - chartStartInset - chartEndInset) ~/ intervalWidth;
+    final remainderContentWidth =
+        contentWidth - (numCompleteIntervals * intervalWidth);
+    final remainderMicros =
+        remainderContentWidth * duration.inMicroseconds / contentWidth;
+    return ((duration.inMicroseconds - remainderMicros) / numCompleteIntervals)
+        .round();
+  }
+
+  int _startingTimestamp(double intervalWidth, int microsPerInterval) {
+    final startingIntervalIndex = horizontalScrollOffset < chartStartInset
+        ? 0
+        : (horizontalScrollOffset - chartStartInset) ~/ intervalWidth + 1;
+    return startingIntervalIndex * microsPerInterval;
+  }
+
+  // TODO(kenz): does this have to return true all the time? Is it cheaper to
+  // compare delegates or to just paint?
+  @override
+  bool shouldRepaint(TimelineGridPainter oldDelegate) => true;
```

Comment:
you should implement equals. Repaint is fairly expensive.

Reply:
done